src/optional.lisp

```diff
@@ -0,0 +1,59 @@
+;;;; optional.lisp
+
+(defpackage #:coalton-impl/optional
+  (:use
+   #:cl)
+  (:export
+   #:cl-none-p
+   #:cl-some-p
+   #:unbox-cl-some))
+
+(in-package #:coalton-impl/optional)
+
+(declaim (inline make-cl-optional))
+(defstruct (cl-optional (:constructor make-cl-optional (depth)))
+  (depth 0 :type (unsigned-byte 16) :read-only t))
+
+(defmethod print-object ((opt cl-optional) stream)
+  (labels ((helper (n)
+             (if (zerop n)
+                 (format nil "~S" 'coalton::None)
+                 (format nil "(~S ~A)" 'coalton:Some (helper (1- n))))))
+    (format stream "#.~A" (helper (cl-optional-depth opt)))))
+
+(defparameter coalton::cl-none (make-cl-optional 0))
+
+(defun cl-none-p (x)
+  (and (typep x 'cl-optional)
+       (zerop (cl-optional-depth x))))
+
+(defparameter cl-optional-cache
+  (make-array 16 :element-type 'cl-optional
+                 :initial-element coalton::cl-none
+                 :fill-pointer 1))
+
+(defun coalton::cl-some (x)
+  (typecase x
+    (cl-optional
+     (let ((new-depth (1+ (cl-optional-depth x))))
```

Comment:
this will simplify to basically

```
(if (< new-depth +optional-cache-size+) (aref +optional-cache+ new-depth) (make-cl-optional new-depth))
```

which should have pretty short disassembly and be OK to inline.

Don't do `length`, `vector-push`, or `ignore-errors` in this crucial function.